File: Data_Structures/Stacks/postfix.py
# Write a program to evaluate a postfix expression using a stack
# The postfix expression will only contain integer operands and the following operators: '+','-','*','/'

class Stack:
    def __init__(self):
        self.stack = []
    
    def is_empty(self):
        return len(self.stack) == 0
    
    def is_full(self):
        return False
    
    def push(self, element):
        self.stack.append(element)
    
    def pop(self):
        if self.is_empty():
            print("Stack is empty. Cannot pop element.")
            return None
        return self.stack.pop()

def evaluate_postfix(postfix):
    stack = Stack()

    for char in postfix:
        if char.isdigit():
            stack.push(int(char))
        else:
            operand2 = stack.pop()
            operand1 = stack.pop()

            if char == '+':
                stack.push(operand1 + operand2)
            elif char == '-':
                stack.push(operand1 - operand2)
            elif char == '*':
                stack.push(operand1 * operand2)
            elif char == '/':
                stack.push(operand1 / operand2)

    return stack.pop()

postfix = "52+83-*4/"

result = evaluate_postfix(postfix)
print("Result:", result)
print("\n")

File: Data_Structures/Stacks/postfix.c
//Write a program to evaluate a postfix expression using a stack
//The postfix expression will only contain integer operands and the following operators: '+','-','*','/'

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#define MAX_SIZE 100

typedef struct {
    int stack[MAX_SIZE];
    int top;
} Stack;

void initialize(Stack* stack) {
    stack->top = -1;
}

int isEmpty(Stack* stack) {
    return stack->top == -1;
}

int isFull(Stack* stack) {
    return stack->top == MAX_SIZE - 1;
}

void push(Stack* stack, int element) {
    if (isFull(stack)) {
        printf("Stack is full. Cannot push element.\n");
        return;
    }
    stack->stack[++stack->top] = element;
}

int pop(Stack* stack) {
    if (isEmpty(stack)) {
        printf("Stack is empty. Cannot pop element.\n");
        return -1;
    }
    return stack->stack[stack->top--];
}

int evaluatePostfix(char* postfix) {
    Stack stack;
    initialize(&stack);
    int len = strlen(postfix);

    for (int i = 0; i < len; i++) {
        if (isdigit(postfix[i])) {
            push(&stack, postfix[i] - '0');
        } else {
            int operand2 = pop(&stack);
            int operand1 = pop(&stack);

            switch (postfix[i]) {
                case '+':
                    push(&stack, operand1 + operand2);
                    break;
                case '-':
                    push(&stack, operand1 - operand2);
                    break;
                case '*':
                    push(&stack, operand1 * operand2);
                    break;
                case '/':
                    push(&stack, operand1 / operand2);
                    break;
            }
        }
    }

    return pop(&stack);
}

int main() {
    char postfix[] = "52+83-*4/";

    int result = evaluatePostfix(postfix);
    printf("Result: %d\n", result);
    printf("\n");

    return 0;
}
